09-graphs-p2/01-graphs-review.md

```diff
@@ -0,0 +1,325 @@
+
+#### Option B
+![Graph B](images/graph_b.png)
+
+#### Option C
+![Graph C](images/graph_c.png)
+
+##### !end-question
+
+##### !options
+
+a| Option A
+b| Option B
+c| Option C
+
+##### !end-options
+
+##### !answer
+
+a|
+
+##### !end-answer
+
+##### !explanation
+
+Option A is correct because the nodes/edges in the adjacency dictionary correspond with the nodes and edges in the graph. 
+
+For instance:
+
+* Seattle is connected to only one city: Portland
+* Portland is connected to Dallas and Seattle
+* Dallas is connected to Boston, Detroit, and Portland
+* Detroit is connected to Dallas
+* Boston is connected to Dallas
+
+##### !end-explanation 
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+
+## Review of Breadth First Search & Depth First Search
+
+### Breadth First Search (BFS)
+You may recall from the last lesson that the Breadth First Search (BFS) algorithm starts with a particular node and then visits each node connected to the starting point before expanding outward.
+
+The BFS algorithm requires us to use a queue to visit nodes in a "level-by-level" ordering by adding each of the neighbors of the starting node to the queue. We then loop through the queue, adding unvisited neighbors to the list of visited nodes as well as to the queue to be further processed.
+
+### Depth First Search (DFS)
+You may also recall from the last lesson that the Depth First Search (DFS) algorithm starts at a node and then follows a path as deeply as possible before backing up and following the next path. We refer to this algorithm as a back-tracking algorithm.
+
+The DFS algorithm requires the use of a stack. The stack is used to further process nodes along the path, starting with the first node and then adding its neighbors to the stack before iterating on the nodes added to the stack. 
+
+### Why DFS does not find shortest path
+
+The main reason the DFS algorithm is not typically used to find the shortest path is because DFS is not a "greedy" algorithm. The algorithm is not designed to account for changes in logic based on the data it encounters, which is what is needed for finding the shortest path in a graph.
+
+## !callout
+A greedy algorithm is an approach for solving a problem by selecting the best option available at the moment. A greedy algorithm never reverses the earlier decision even if the choise is wrong. An algorithm which is considered greedy may not produce the best result for all problems.
+## !end-callout
+
+### Why BFS (kinda) allows for shortest path of an unweighted graph
+
+By itself, BFS does *not* allow for finding the shortest path of an unweighted graph. In order to retrieve the shortest path using the BFS algorithm, one would need to modify the algorithm to store the current shortest distance to the target node as well as the preceding node in the shortest path. This is essentially what Djikstra's algorithm does, which we will learn more about in the next section.
```

Comment:
I like calling out that the version of BFS we had them implement would not allow them to find the shortest path, we would need to store preceding nodes to allow for backtracking.

However, it might be helpful to expand more upon how BFS operates - it visits nodes based on proximity. It first visits all neighbors, aka all nodes that are one edge away from the start node, then the neighbor's neighbors aka all nodes that are two edges away, etc. I personally like returning to the BST to visualize this (a subtype of a graph here) since it is by nature hierarchical and has 'levels'. 

I also like setting them up for Dijkstra's here. It could be helpful to contrast them by clarifying the difference is it uses weights to determine the next node to consider rather than the number of 'hops'/edges.